### PR TITLE
feat: custom priority mempool

### PR DESCRIPTION
**Pull request by skosito**
# Description

In cosmos 0.47, application mempools are possible https://docs.cosmos.network/main/build/building-apps/app-mempool
There are 2 provided implementation, sender nonce and priority nonce mempools. Priority nonce seems like it's most suitable for this usecase - give more priority to system txs.

Mempools from cosmos 0.47 assume signatures can be extracted from sdk.Tx, but it doesn't work for ethermint txs. On latest cosmos-sdk main branch, signer extractor adapter is present, which can be used to provide custom adapter to pull out senders and nonces (check cronos ethermint for example). 
But, on 0.47 this option is not present, so I forked priority nonce mempool, and default proposal handler, and only replaced part with that custom signer adapter.

Tests will be added in:
https://github.com/zeta-chain/node/issues/2182
https://github.com/zeta-chain/node/issues/2183

Closes: https://github.com/zeta-chain/node/issues/734

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Include instructions and any relevant details so others can reproduce. 

- [ ] Tested CCTX in localnet
- [ ] Tested in development environment
- [ ] Go unit tests
- [ ] Go integration tests
- [ ] Tested via GitHub Actions 

# Checklist:

- [ ] I have added unit tests that prove my fix feature works

### CODE REVIEW

File: app/mempool/senders_with_nonce.go [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	authante "github.com/cosmos/cosmos-sdk/x/auth/ante"
+	"github.com/cosmos/cosmos-sdk/x/auth/signing"
+	evmtypes "github.com/evmos/ethermint/x/evm/types"
+)
+
+func GetSendersWithNonce(tx sdk.Tx) ([]SenderWithNonce, error) {
+	if txWithExtensions, ok := tx.(authante.HasExtensionOptionsTx); ok {
+		opts := txWithExtensions.GetExtensionOptions()
+		if len(opts) > 0 && opts[0].GetTypeUrl() == "/ethermint.evm.v1.ExtensionOptionsEthereumTx" {
```
Comment (lumtis) [inline app/mempool/senders_with_nonce.go:15]: NIT: Not possible to get the URL from the type directly? Or maybe use a constant

File: app/mempool/senders_with_nonce.go [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+import (
+	"fmt"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	authante "github.com/cosmos/cosmos-sdk/x/auth/ante"
+	"github.com/cosmos/cosmos-sdk/x/auth/signing"
+	evmtypes "github.com/evmos/ethermint/x/evm/types"
+)
+
+func GetSendersWithNonce(tx sdk.Tx) ([]SenderWithNonce, error) {
```
Comment (lumtis) [inline app/mempool/senders_with_nonce.go:12]: Let's add comment to the function
We can describe why needed, we compute the signer differently for evm messages?

File: app/mempool/priority_nonce_mempool.go [hunk 3]
```diff
@@ -0,0 +1,448 @@
+// This proposal handler is taken from https://github.com/cosmos/cosmos-sdk/blob/v0.47.10/types/mempool/priority_nonce.go
+// Only difference is extraction of senders and nonce from tx. In latest version of cosmos, there is a way to provide adapter for this, but in 0.47.10 this is the only way.
+// TODO: remove this once we upgrade cosmos
```
Comment (lumtis) [inline app/mempool/priority_nonce_mempool.go:3]: Can we create an issue and give link to it here?

File: app/mempool/priority_nonce_mempool.go [hunk 4]
```diff
@@ -0,0 +1,448 @@
...
+	return nil
+}
+
+func IsEmpty(mempool mempool.Mempool) error {
+	mp := mempool.(*PriorityNonceMempool)
+	if mp.priorityIndex.Len() != 0 {
+		return fmt.Errorf("priorityIndex not empty")
+	}
+
+	var countKeys []int64
```
Comment (lumtis) [inline app/mempool/priority_nonce_mempool.go:425]: I think we can fix the lint here even if it was forked

File: app/mempool/senders_with_nonce.go [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+	return getSendersWithNonceDefault(tx)
+}
+
+type SenderWithNonce struct {
+	Sender string
+	Nonce  uint64
+}
+
+func getSendersWithNonceDefault(tx sdk.Tx) ([]SenderWithNonce, error) {
```
Comment (lumtis) [inline app/mempool/senders_with_nonce.go:38]: Need comment to describe what default mean in this context

File: app/ante/priority.go [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+}
+
+// AnteHandle implements AnteDecorator
+func (vad SystemPriorityDecorator) AnteHandle(
+	ctx sdk.Context,
+	tx sdk.Tx,
+	simulate bool,
+	next sdk.AnteHandler,
+) (sdk.Context, error) {
+	newCtx := ctx.WithPriority(math.MaxInt64)
```
Comment (lumtis) [inline app/ante/priority.go:27]: Don't we need to filter if the tx contains only system messages here?

**Review by lumtis — COMMENTED** [review]
Nice.

Any way it has been tested? I may think we could maybe introduce a new local env setup where the max gas per block is low and we send many txs, checking if system txs pass through.
This can be a general setup for stress tests, we currently have a stress tests command but it will need to be improved

File: app/mempool/senders_with_nonce.go [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+	evmtypes "github.com/evmos/ethermint/x/evm/types"
+)
+
+// GetSendersWithNonce is used to extract sender and nonce information txs
+// if tx is ethermint, it is extracted using from and nonce field
+// if it's cosmos tx, default cosmos way using signatures is used
+func GetSendersWithNonce(tx sdk.Tx) ([]SenderWithNonce, error) {
+	const extensionOptionsEthereumTxTypeURL = "/ethermint.evm.v1.ExtensionOptionsEthereumTx"
+	if txWithExtensions, ok := tx.(authante.HasExtensionOptionsTx); ok {
+		opts := txWithExtensions.GetExtensionOptions()
```
Comment (kingpinXD) [inline app/mempool/senders_with_nonce.go:20]: Nit: but it makes it more readable if we have two functions 
`getSendersWithNonceEthermint`  && `getSendersWithNonceCosmos` .

**Review by kingpinXD — APPROVED** [review]

**Review by lumtis — APPROVED** [review]
>@lumtis thanks for review, fixed the comments
regarding testing, i think for unit tests we can also fork them and add ethermint handling if possible in this PR, will do that here
regarding e2e and stress tests, maybe we can discuss proper setup and handle in follow up PR?

Sounds good to me, do we already have a follow up issue?

**Review by gartnera — APPROVED** [review]
stamp

**Pull request merged**